### PR TITLE
test: Remove binary manifest list avro file

### LINKED ISSUE #70: test: Replace binary avro file by generating it on the fly.
I think it is worth the effort to generate the files. This way we can also generate the whole structure later on (manifest-list, manifest, datafile). Since Iceberg requires absolute paths, it is not easy to generate these files on forehand (looking at the Windows CI 😁)

_Originally posted by @Fokko in https://github.com/apache/iceberg-rust/pull/56#discussion_r1338197094_

### PR DESCRIPTION
**Pull request by liurenjie1024**
Close #70 

We generate manifest list avro file on the fly.

### CODE REVIEW

**Review by Xuanwo — APPROVED** [review]
Great!

File: crates/iceberg/src/spec/manifest_list.rs [hunk 1]
```diff
@@ -931,55 +931,104 @@ mod test {
...
+            ]
+        };
+
+        let file_io = FileIOBuilder::new_fs_io().build().unwrap();
+
+        let tmp_dir = TempDir::new().unwrap();
+        let file_name = "simple_manifest_list_v1.avro";
+        let full_path = format!("{}/{}", tmp_dir.path().to_str().unwrap(), file_name);
+
+        let mut writer = ManifestListWriter::v1(
```
Comment (Xuanwo) [inline crates/iceberg/src/spec/manifest_list.rs:964]: I'm bit worried about this. What will happen if we have bugs in both `ManifestListWriter` and `ManifestListReader`?

Comment (liurenjie1024) [reply crates/iceberg/src/spec/manifest_list.rs:964]: Yes, that's a problem. But I don't see other ways to do it. I think such kinds of bugs can only be avoid during integrations tests against other systems such as spark.

Comment (ZENOTME) [reply crates/iceberg/src/spec/manifest_list.rs:964]: Can we call FastAvro(python) in rust to generate the file for testing🤔?

Comment (liurenjie1024) [reply crates/iceberg/src/spec/manifest_list.rs:964]: It's possible. But I took a look at python implementation, it seems that they implemented the encoding/decoding of avro in pyiceberg, that's why they need cross validation of another library. This is not the case for rust. I think if we want to verify correctness, maybe setting up integration tests against systems like spark would be more appropriate.

Comment (ZENOTME) [reply crates/iceberg/src/spec/manifest_list.rs:1080]: I find that avro-tool can generate an avro file by schema file and data JSON file. https://www.michael-noll.com/blog/2013/03/17/reading-and-writing-avro-files-from-the-command-line/

Comment (liurenjie1024) [reply crates/iceberg/src/spec/manifest_list.rs:1080]: Interesting, it seems that it requires java to run.

Comment (ZENOTME) [reply crates/iceberg/src/spec/manifest_list.rs:1080]: > Interesting, it seems that it requires java to run. 

Yes. Another way is to run it in docker https://hub.docker.com/r/coderfi/avro-tools.🙈

Comment (liurenjie1024) [reply crates/iceberg/src/spec/manifest_list.rs:1080]: I did some tests and found that the json file doesn't contains all information, for example meta is missing. The docker image is not well maintained, for example last update time is 8 years ago. Given the complexity introduced, I think using integration tests with other systems maybe better?

Comment (ZENOTME) [reply crates/iceberg/src/spec/manifest_list.rs:1080]: > I did some tests and found that the json file doesn't contains all information, for example meta is missing. The docker image is not well maintained, for example last update time is 8 years ago. Given the complexity introduced, I think using integration tests with other systems maybe better?

Thanks for your test! So seems there is no better way. Using integration tests with other systems also looks good to me. The unit test can be used to test that the reader and writer can process each other.

Comment (liurenjie1024) [reply crates/iceberg/src/spec/manifest_list.rs:1080]: Yes, keeping unit tests simple and easy to maintain. We can leave parts difficult to setup to integration tests.😁

**Review by ZENOTME — APPROVED** [review]
Thanks!

**Review by Xuanwo — APPROVED** [review]
Let's move!

File: crates/iceberg/src/spec/manifest.rs [hunk 2]
```diff
@@ -1311,511 +1311,452 @@ mod tests {
...
+                },
+                content: ManifestContentType::Data,
+                format_version: FormatVersion::V2,
+            },
+            entries: vec![
+                ManifestEntry {
+                    status: ManifestStatus::Added,
+                    snapshot_id: Some(0),
+                    sequence_number: Some(1),
+                    file_sequence_number: Some(1),
```
Comment (Fokko) [inline crates/iceberg/src/spec/manifest.rs:1395]: Typically when a new manifest-entry is being added, the snapshot-id and sequence-numbers are not set and inherited from the snapshot itself.

**Review by Fokko — APPROVED** [review]
One small nit, great to get rid of those binaries! Thanks @liurenjie1024

File: crates/iceberg/src/spec/manifest_list.rs (hunk 1 above)
Comment (Fokko) [reply crates/iceberg/src/spec/manifest_list.rs:1080]: I have the same concern as well. Would be great to have some additional tests around Iceberg-specific things, for example, if the field-id resolution works correctly.

File: Makefile [hunk 3]
```diff
@@ -34,6 +34,9 @@ cargo-sort:
 
 check: check-fmt check-clippy cargo-sort
 
+unit-test:
```
Comment (Fokko) [inline Makefile:37]: Are these ran in the CI?

Comment (liurenjie1024) [reply Makefile:37]: The ci will run all tests, using the `make test` command. This command just makes running unit test convenient.

File: crates/iceberg/src/spec/manifest.rs (hunk 2 above)
Comment (liurenjie1024) [reply crates/iceberg/src/spec/manifest.rs]: Fixed. In v2 by default it's `None`, and in v1 it should be `0`.

**Comment by Fokko** [comment]
Thanks again @liurenjie1024 for working on this. @Xuanwo and @ZENOTME thanks for the review 👍

**Pull request merged**